File: category/dp/11053.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	var n int
	var dp [1001]int
	var numlist []int

	reader := bufio.NewReader(os.Stdin)
	fmt.Fscan(reader, &n)
	numlist = make([]int, n)

	for i := 0; i < n; i++ {
		fmt.Fscan(reader, &numlist[i])
	}

	var ans int = 0
	for i := 1; i < n; i++ {
		num1 := numlist[i]
		for j := i - 1; j >= 0; j-- {
			num2 := numlist[j]

			if num2 < num1 {

				if dp[i] > dp[j] {
					continue
				}
				dp[i] = dp[j] + 1

				if ans < dp[i] {
					ans = dp[i]
				}
			}
		}
	}

	fmt.Printf("%d \n", ans+1)
}

File: category/dp/11053.py
import sys

def main():
    n = int(input())
    dp = [0] * 1001
    # numlist = []

    # arr = list(map(int, sys.stdin.readline().split()))
    # numlist = list(map(int, sys.stdin.readline().split()))
    # arr = list(map(int, input().split()))
    numlist = list(map(int, input().split()))

    ans = 0

    for i in range(1, n):
        num1 = numlist[i]
        for j in range(i-1, -1, -1):
            num2 = numlist[j]

            if num2 < num1:
                if dp[i] > dp[j]:
                    continue
                dp[i] = dp[j] + 1
                if ans < dp[i]:
                    ans=dp[i]


    print(ans+1)


if __name__ == "__main__":
    main()